### ISSUE TITLE
maya plugin: support for stripping namespaces in "usdExport"

### ISSUE THREAD

**Issue by nxkb** [PixarAnimationStudios/USD#126]
Hello! I wanted to poke in on this issue which will impact a lot of people's Maya integration and guage the interest in a pull request to address it. 
### Description of Issue
When exporting usd out of maya, we will need to be able to strip a variable number of namespaces on referenced nodes. Currently they become part of the prim name, which makes it problematic later on to figure out whether they were originally namespaces or actually part of the node names. I believe this feature is pretty critical for studios that are doing animation in maya...

If you start with a referenced rigg (that also contains another reference), this is your hierarchy in maya:
```
wolf1:werewolf_top
 - wolf1:def:werewolf_mesh
```
and you export usd:
```
 wolf1_werewolf_top
 - wolf1_def_werewolf_mesh
```
This makes things tricky... later operations like name based shader assignment will have to work around this extra prefix on _every_ prim. So to match the node, you would need a rule like:
```*/*_werewolf_top/*_def_werewolf_mesh```

However, this would be the hierarchy with support for stripping (1) level of namespaces:
```
 werewolf_top
 - def_werewolf_mesh
```
which means you could have a much simpler expression to match the same node (for all your werewolves!):
```*/werewolf_top/def_werewolf_mesh```

One approach to solve this would be to add a custom export Chaser that adjusts exporting node names to strip a certain number of namespaces from the prim name. However, this behavior may be common enough to warrant something more built-in than just a custom Chaser. I'm sure many users would like to skip the setup of a chaser to get this behavior. 

So I'm wondering:
Are there any efforts already under way to handle this issue? Would you suggest adding support to the maya plugin directly or going through the chaser plugin system? Would we be able to contribute this back to the USD repo? 

Thanks!
-- nick kb

**Comment by jtran56** [comment]
Filed as internal issue #141438.

**Comment by spiffmon** [comment]
Hi Nick,
   We'd like to first investigate whether we can have full roundtripping support for Maya namespaces, i.e. something like, preserving in USD prim metadata, on **export**, the actual Maya nodename and namespaces; then on **import**, for any prim that has that metadata, record on the Maya node the USD prim name, which would in turn be used on re-export rather than re-computing a new USD name from the Maya name (pros and cons of that behavior... not sure... but the initial export/import seems reasonable).

Would that address your needs, or would there still be a need/desire for selective stripping of Maya namespaces?

**Comment by nxkb** [comment]
Hi Spiff,
I like the way that a metadata approach would store the original state non-destructively, but I could see it run into problems if there were name clashes between separate namespaces. How would you name the Usd Prims in the following export? 
```
flock_rig: top_xform
- flock_rig:goose1:goose_head
- flock_rig:goose2:goose_head
```
Unless we can come up with a clever solution to handle this case, it seems like it could be messier than simply stripping 1 level of namespaces. (We currently export animation on a rig reference, and strip a single namespace so that names will be consistent for that rig no matter what namespace it was referenced with. Within the rig, namespaces are still useful to have multiple copies of a subreference like in the above example.)

**Comment by spiffmon** [comment]
Hi Nick,
   Sorry I wasn't clear: the metadata would be a pure addition to what we are currently doing in export, for exactly the reasin you provided - if Maya is using namespaces to create uniquely-named siblings that otherwise would not be unique, then we need to preserve that in our prim names.  So the USD would look the same as it does today, but we'd add new logic to the import that uses the metadata, when present, anytime it needs to translate a USD path to a Maya dagPath.

   I don't know much about the different ways/purposes namespaces are used in Maya, but it seems like there could be legal Maya files that would fail to roundtrip cleanly with a fixed rule like strip a single level of namespace on export?

**Comment by chadrik** [comment]
Hi @spiffmon 
I believe that these are two independent features.  Even with the ability to round-trip namespaces, it is still often necessary to strip N levels of namespaces on export.  Currently we use Alembic, which supports maya-style namespaces on node names, but we still use the `-stripNamespaces 1` option during export.  

In the case our our USD pipeline, the maya namespace will be a "workarea" where animation is added to a Maya rig, since namespaces are Maya's only means of uniqifying names, but downstream we will use USD references. 

Here's a more elaborate example based on Nick's.

Animator loads an initial scene:

```
world
  chars
    wolf1  <-- wolf.usd (placeholder for maya rig wolf.mb)
    wolf2  <-- wolf.usd (placeholder for maya rig wolf.mb)
```

This would trigger the creation of references where animators will apply their animation:

```
world
  chars
    wolf1  <-- wolf.usd
    wolf2  <-- wolf.usd
workarea
  wolf1:werewolf_top  <-- wolf.mb
    wolf1:def:werewolf_mesh
      wolf1:...
  wolf2:werewolf_top  <-- wolf.mb
    wolf2:def:werewolf_mesh
      wolf2:...
```

On export of our werewolves, we create a variant that overrides the placeholders with our newly authored animation caches.  We definitely do not want the wolf1_ and wolf2_ prefixes within our exports since that just complicates things downstream, as Nick explained.

After stripping off one level of namespaces, our exports look like this:

```
werewolf_top
  def_werewolf_mesh
    ...
```

And the resulting USD scene looks like this:

```
world
  chars
    wolf1  <-- wolf1_anim.usd
      def_werewolf_mesh
        ...
    wolf2  <-- wolf2_anim.usd
      def_werewolf_mesh
        ...
```

Whether `def_` should be read back into maya as `def:` is a separate issue and one that's much less important to us at the moment.

**Comment by spiffmon** [comment]
I think I see, @chadrik .  So, the root prim of both wolf.usd and wolf1_anim.usd and wolf2_anim.usd  (those last two you get by doing two selection-based exports of the same Maya scene, if it looked like your original example?)  would be </werewolf_top> ?

Apologies to @nxkb for not realizing namespace-stripping was an already-supported feature of the alembic exporter, and it seems important, as you both describe, for Maya rigging and animation workflows.

My remaining questions (for now!) are about how ns-stripping should interact with re-import into Maya (or also useful would be to know if importing animated scenes into Maya with the same namespace topology they had during animation is actually not a useful problem to solve).

1. If you strip N namespaces on export, and that causes two sibling trees of your Maya scene to collide in "USD namespace topology", what do we do?  That would happen in @chadrik 's example if you exported the whole scene at once.
2. If we preserve "originating Maya namespaces" in metadata on export, should the stripped namespaces be included?
3. Depending on the answer to 2, do we include the stripped namespaces when reconstituting the Maya scene on import of the USD scene?

**Comment by chadrik** [comment]
> So, the root prim of both wolf.usd and wolf1_anim.usd and wolf2_anim.usd would be </werewolf_top> ?

correct, but that idea is about to change based on the plan below.

> My remaining questions (for now!) are about how ns-stripping should interact with re-import into Maya.

Your questions have prompted a few more ideas, so brace yourself for excessive detail :)

In my first proposal I was thinking about the bare minimum features required to bolt together a Maya referencing system custom tailored to our pipeline.  Below is a proposal for a more generic and robust referencing/namespace workflow which would (hopefully) work out of the box, but still have enough flexibility to meet the demands of most pipelines.

The examples below begin with the original scene described in my previous post, repasted here:

```
world <-- shot.usd
  chars
    wolf1  <-- wolf.usd (placeholder for maya rig wolf.mb)
      ...
    wolf2  <-- wolf.usd (placeholder for maya rig wolf.mb)
      ...
```

## Import

We would add a mechanism within the USD Maya plugin for triggering the creation of Maya references from placeholder prims (by watching for prims with special `maya:reference` attribute, or something similar, as discussed on usd-interest).  

Maya references would be added in the following fashion:

- **placed under a group (i.e. transform) named after the reference prim**: the group is required because Maya references can contain many top-level nodes, and there is no concept akin to a default prim: i.e. no facility for specifying a default node, renaming nodes, or filtering nodes.  So we imagine that the root of the referenced scene is the default prim.  (Originally I had planned to have our pipeline enforce a single top-level node, but I think using a group in lieu of that requirement is more robust).
- **namespaces for the reference could be created in one of 4 ways** (configurable per reference using additional `maya:reference_namespace_mode` attribute?):
  - A) **no namespace**: chance of clashes for DG nodes is high for multiple references, but could be a nice option if you *know* there are no DG nodes in your Maya scenes (or don't care if they get renamed).
  - B) **use reference prim name** (e.g. `wolf1:`): 
    - with merge namespaces enabled: clashes will occur for DG nodes if two references are loaded into the same namespace (e.g. if `/group1/wolf1` and `/group2/wolf1` are both maya references). 
    - with merge namespaces disabled:  if two references are loaded into the same namespace, the second one will be auto-assigned a uniquified namespace. clash-free, but with side-effects.
  - C) **use reference prim path** (e.g. `chars:wolf1:`): clash-free.
  - D) **custom string**:  up to pipeline to ensure safety

(Note that the namespaces are required only for DG nodes, since unlike USD, Maya's DG nodes do not have a parent DAG node and therefore cannot derive any namespace scoping from DAG paths.)

After loading the scene using option B:

```
world <-- shot.usd
  chars
    wolf1  <== wolf.mb (reference is parented below here)
      wolf1:werewolf_top  
        wolf1:def:werewolf_mesh
          wolf1:...
    wolf2  <== wolf.mb (reference is parented below here)
      wolf2:werewolf_top
        wolf2:def:werewolf_mesh
          wolf2:...
```

Or after option C:

```
world <-- shot.usd
  chars
    wolf1  <== wolf.mb (reference is parented below here)
      chars:wolf1:werewolf_top  
        chars:wolf1:def:werewolf_mesh
          chars:wolf1:...
    wolf2  <== wolf.mb (reference is parented below here)
      chars:wolf2:werewolf_top
        chars:wolf2:def:werewolf_mesh
          chars:wolf2:...
```

I'd probably start by implementing Option C, and add the others as requested.  It's heinously ugly, but safest.

## Export 

On export we drop any namespaces added by the import mechanism.

In our example, we would produce the following USD scene:

```
chars
  wolf1
    werewolf_top
      def_werewolf_mesh
        ...
  wolf2
    werewolf_top
      def_werewolf_mesh
        ...
```

## Re-Import

When re-importing exported USD scenes back into Maya, we could provide several options for namespace handling (assuming we author the appropriate metadata on export):
- use USD names. i.e. continue to do nothing.
- recreate namespaces which were not created by the import system.  e.g. if we originally imported with option C (above):

  ```
  world <-- shot.usd
    chars
      wolf1
        werewolf_top  
          def:werewolf_mesh
            ...
      wolf2
        werewolf_top
          def:werewolf_mesh
            ...
  ```
- recreate the original namespaces for all nodes.  e.g. if we originally imported with option C (above):
  ```
  world <-- shot.usd
    chars
      wolf1
        chars:wolf1:werewolf_top  
          chars:wolf1:def:werewolf_mesh
            chars:wolf1:...
      wolf2
        chars:wolf2:werewolf_top
          chars:wolf2:def:werewolf_mesh
            chars:wolf2:...
  ```

## Summary

Let me summarize a few high level concepts:

- we create a system, incorporated into the Maya plugin, which treats Maya's references+namespaces as an analog to USD's references.
- "native" namespaces created by the system within Maya are never exported to USD.  Doing so would be redundant as we're treating them as analogues.  (`-stripNamespace` as an export flag is therefore not necessary)
- "foreign" namespaces created outside the system (e.g. by nested Maya references) continue to be converted to valid USD paths by replacing the colon `:`  
- on re-import, foreign and native namespaces can be optionally recreated (by using metadata injected into exported USD files)

@spiffmon , your questions made me realize that if the plugin controls things on the import side, we can side-step certain pitfalls inherent in the free-style approach that I originally proposed.  This new proposal is more automated and safer, with the trade-off that it is more rigid.  Hopefully, the various options that I mentioned will counter-balance the rigidity.

Thoughts? 

One last question: @spiffmon , can you provide a regex or a description of a valid prim name in USD?  Perhaps if there is a character that is valid in USD but not in Maya, we can use this to replace the colon instead of an underscore on export, and thereby avoid potential name clashes (e.g. between nodes `foo:bar` and `foo_bar`)

**Comment by gitamohr** [comment]
> One last question: @spiffmon , can you provide a regex or a description of a valid prim name in USD?

Valid prim names are C/Python identifiers, so `[_a-zA-Z][_a-zA-Z0-9]*`

**Comment by nxkb** [comment]
I've added PR https://github.com/PixarAnimationStudios/USD/pull/327 that implements a scaled back version of namespace stripping. Our needs have evolved and become a lot simpler since we had this discussion, so that PR is enough for us at the moment. It could also be used as a base to implement further support.

**Comment by sunyab** [comment]
PR #327 was merged and included in release 18.09. Closing, thanks!
